src/main/kotlin/org/rust/lang/core/psi/ext/RsPath.kt

```diff
@@ -9,20 +9,45 @@ import com.intellij.lang.ASTNode
 import com.intellij.psi.PsiElement
 import com.intellij.psi.stubs.IStubElementType
 import org.rust.lang.core.macros.RsExpandedElement
+import org.rust.lang.core.psi.RsElementTypes.*
 import org.rust.lang.core.psi.RsPath
+import org.rust.lang.core.psi.tokenSetOf
 import org.rust.lang.core.resolve.ref.RsPathReference
 import org.rust.lang.core.resolve.ref.RsPathReferenceImpl
 import org.rust.lang.core.stubs.RsPathStub
 
+private val RS_PATH_KINDS = tokenSetOf(IDENTIFIER, SELF, SUPER, CSELF, CRATE)
+
 val RsPath.hasColonColon: Boolean get() = stub?.hasColonColon ?: (coloncolon != null)
-val RsPath.hasCself: Boolean get() = stub?.hasCself ?: (cself != null)
+val RsPath.hasCself: Boolean get() = kind == PathKind.CSELF
+val RsPath.kind: PathKind get() {
+    val stub = stub
+    if (stub != null) return stub.kind
+    val child = node.findChildByType(RS_PATH_KINDS)
+    return when (child?.elementType) {
+        IDENTIFIER -> PathKind.IDENTIFIER
+        SELF -> PathKind.SELF
+        SUPER -> PathKind.SUPER
+        CSELF -> PathKind.CSELF
+        CRATE -> PathKind.CRATE
+        else -> error("Malformed RsPath element: `$text`")
+    }
+}
```

Comment:
I don't remember exactly, but if `<Type as Trait>::AssocType` thingies are also paths, this might explode?

Reply:
Current grammar rule for this case is `Path ::= (Path '::' | '::' | TypeQual)? PathIdent PathTypeArguments?`

So, in this case `AssocType` is an identifie, `<Type as Trait>` is a `TypeQual` and this code returns `PathKind.IDENTIFIER`